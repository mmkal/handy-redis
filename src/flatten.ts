export const flattenDeep = (array: any[]): any[]  => {
    const flat = [] as any[];
    for (const item of array) {
        Array.isArray(item) ? flat.push(...flattenDeep(item)) : flat.push(item);
    }
    return flat;
};
