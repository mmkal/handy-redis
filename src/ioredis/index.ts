import * as IORedis from "ioredis";
import { Commands } from "../generated/interface";
import { WrappedIORedisMulti } from "./multi";

export interface WrappedIORedisClient extends Omit<Commands, "multi"> {
    ioredis: IORedis.Redis;
    multi(): WrappedIORedisMulti;
}

export interface CreateIORedisClient {
    (): WrappedIORedisClient;
    (ioredis: IORedis.Redis): WrappedIORedisClient;
    (port?: number, host?: string, options?: IORedis.RedisOptions): WrappedIORedisClient;
    (host?: string, options?: IORedis.RedisOptions): WrappedIORedisClient;
    (options?: IORedis.RedisOptions): WrappedIORedisClient;
}

export const createIORedisClient: CreateIORedisClient = (...clientArgs: any[]): WrappedIORedisClient => {
    const ioredis = clientArgs[0] instanceof IORedis ? clientArgs[0] : new IORedis(...clientArgs);

    return Object.assign(ioredis, { ioredis }) as any;
};
