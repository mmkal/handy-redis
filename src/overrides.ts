import { IHandyRedis } from "./generated/interface";
import { Multi } from "redis";

export const useUnderlyingImpl = new Set<keyof IHandyRedis>([
    "multi"
]);

export const additionalFunctions = {
    /** promisified multi execution */
    execMulti: <T = {}>(multi: Multi) => new Promise<T[]>(
        (resolve, reject) => multi.exec((err, data) => err ? reject(err) : resolve(data))
    ),
};

export type AdditionalFunctions = typeof additionalFunctions;
