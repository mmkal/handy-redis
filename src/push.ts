// Note: this is replaced with ./push.ts34.ts for typescript <4
export type Push<A extends unknown[], B> = [...A, B];
