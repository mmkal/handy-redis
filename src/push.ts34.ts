// Typescript <4 doesn't support ... at the beginning of tuples, so this just creates an array with a union of all types
export type Push<A extends unknown[], B> = Array<A[number] | B>;
