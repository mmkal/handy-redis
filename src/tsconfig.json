{
    "extends": "../tsconfig.json",
    "compilerOptions": {
        "sourceMap": false,
        "declaration": true,
        "outDir": "../dist"
    }
}
