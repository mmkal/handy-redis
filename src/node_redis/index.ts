import * as nodeRedis from "redis";
import { Commands } from "../generated/interface";
import { flattenDeep } from "../flatten";
import { WrappedNodeRedisMulti, WrappedNodeRedisMultiImpl } from "./multi";

export interface WrappedNodeRedisClient extends Omit<Commands, "end" | "multi"> {
    nodeRedis: nodeRedis.RedisClient;
    /** @deprecated use `nodeRedis` */
    redis: nodeRedis.RedisClient;

    /** execMulti was removed in v2.x. Use `multi().exec()` instead. */
    readonly execMulti: never;

    multi(): WrappedNodeRedisMulti;
    batch(): WrappedNodeRedisMulti;
    end(flush?: boolean): void;
}

const WrappedNodeRedisClientImpl = class _WrappedNodeRedisClient {
    readonly nodeRedis: nodeRedis.RedisClient;

    /** @deprecated use `nodeRedis` */
    readonly redis: nodeRedis.RedisClient;

    /** execMulti was removed in v2.x. Use `multi().exec()` instead. */
    readonly execMulti: never;

    constructor(nodeRedis: nodeRedis.RedisClient) {
        this.nodeRedis = nodeRedis;
        this.redis = nodeRedis;
    }

    multi(): WrappedNodeRedisMulti {
        return WrappedNodeRedisMultiImpl.create(this.nodeRedis.multi());
    }

    batch(): WrappedNodeRedisMulti {
        return WrappedNodeRedisMultiImpl.create(this.nodeRedis.batch());
    }

    end(flush?: boolean) {
        return this.nodeRedis.end(flush);
    }
};

const addCommands = (methods: string[]) => {
    methods.forEach(method => {
        (WrappedNodeRedisClientImpl.prototype as any)[method] = function (...args: any[]) {
            return new Promise((resolve, reject) => {
                const flattenedArgs = [
                    ...flattenDeep(args),
                    (err: any, reply: any) => (err ? reject(err) : resolve(reply)),
                ];
                return this.nodeRedis[method](...flattenedArgs);
            });
        };
    });
};

addCommands(
    Object.keys(nodeRedis.RedisClient.prototype).filter(
        method => method === method.toLowerCase() && !(method in WrappedNodeRedisClientImpl.prototype)
    )
);

export interface CreateNodeRedisClient {
    (port_arg: number, host_arg?: string, options?: nodeRedis.ClientOpts): WrappedNodeRedisClient;
    (unix_socket: string, options?: nodeRedis.ClientOpts): WrappedNodeRedisClient;
    (options?: nodeRedis.ClientOpts): WrappedNodeRedisClient;
    (redisClient: nodeRedis.RedisClient): WrappedNodeRedisClient;
}

/** add a command that isn't included by default in node_redis e.g. `addCommand('lpos')` */
export const addNodeRedisCommand = (command: string) => {
    nodeRedis.addCommand(command);
    addCommands([command]);
};

export const createNodeRedisClient: CreateNodeRedisClient = (...clientArgs: any[]): WrappedNodeRedisClient => {
    const nodeRedisInstance =
        typeof clientArgs[0] === "object" && typeof clientArgs[0].scan === "function"
            ? clientArgs[0]
            : nodeRedis.createClient.apply(null, clientArgs);

    return new WrappedNodeRedisClientImpl(nodeRedisInstance) as any;
};
