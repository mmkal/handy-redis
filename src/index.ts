import { createNodeRedisClient, WrappedNodeRedisClient } from "./node_redis";

export {
    addNodeRedisCommand,
    createNodeRedisClient,
    WrappedNodeRedisClient,
    CreateNodeRedisClient,
} from "./node_redis";

export { createIORedisClient, WrappedIORedisClient } from "./ioredis";

// aliases for backwards-compatibility with v1.x

/** @deprecated use `createNodeRedisClient` */
export const createHandyClient = createNodeRedisClient;

/** @deprecated use `WrappedNodeRedisClient` */
export type IHandyRedis = WrappedNodeRedisClient;
