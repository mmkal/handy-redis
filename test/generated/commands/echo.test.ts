import { zip, padEnd } from "lodash";
import { IHandyRedis, createHandyClient } from "../../../src";
import { getOverride } from "../../_manual-overrides";
let handy: IHandyRedis;
beforeAll(async () => {
    handy = createHandyClient();
    await handy.ping("ping");
});
beforeEach(async () => {
    await handy.flushall();
});

it("scripts/redis-doc/commands/echo.md example 1", async () => {
    const overrider = getOverride("scripts/redis-doc/commands/echo.md");
    let snapshot: any;
    const commands = [
        `await handy.echo("Hello World!")`,
    ];
    const output: any[] = [];
    try {
        output.push(await handy.echo("Hello World!"));
        const overridenOutput = overrider(output);
        snapshot = zip(commands, overridenOutput).map(pair => `${padEnd(pair[0], 33)} => ${JSON.stringify(pair[1])}`);
    } catch (err) {
        snapshot = { _commands: commands, _output: output, err };
    }
    expect(snapshot).toMatchSnapshot();
});
