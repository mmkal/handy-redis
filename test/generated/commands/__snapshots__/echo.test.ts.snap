// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scripts/redis-doc/commands/echo.md example 1 1`] = `
Array [
  "await handy.echo(\\"Hello World!\\")  => \\"Hello World!\\"",
]
`;
